Fixes #35593 - Use CentOS 8 Stream for container

ekohl opened the pull request:
Currently the container image is based on Fedora 33, but that's long EOL. Current Fedora versions have Ruby 3.1, but we're unable to use that. Changing it to CentOS 8 Stream is the closest match and also close to what we run in production.

It also updates to PostgreSQL 12 which we use in production.

To make docker-compose.yml compatible with podman-compose the port is changed to a string.

It should be noted that since 36be84a1cfae3c207f05975ba4cac8f448495a6f no assets are served anymore so effectively the UI can no longer be used. Additionally, since 20f4804784c622851fd7beade233797c95b1304b websockify is unbundled and the container now no longer has it. Hence why it's a draft.

ekohl committed:
Fixes #35593 - Use CentOS 8 Stream for container

Currently the container image is based on Fedora 33, but that's long
EOL. Current Fedora versions have Ruby 3.1, but we're unable to use
that. Changing it to CentOS 8 Stream is the closest match and also close
to what we run in production.

It also updates to PostgreSQL 12 which we use in production.

To make docker-compose.yml compatible with podman-compose the port is
changed to a string.

It should be noted that since 36be84a1cfae3c207f05975ba4cac8f448495a6f
no assets are served anymore so effectively the UI can no longer be
used. It now uses the RAILS_SERVE_STATIC_FILES env var that the
container already set.

ekohl commented:
@evgeni I recall us talking about this the other day.

adamruzicka commented:
What remains to be done here to undraft this?

ekohl commented:
I suppose finding a way to serve assets again. Perhaps it's as simple as adding a sed to replace false with true in `config/environments/production.rb`.

evgeni commented:
> I suppose finding a way to serve assets again. Perhaps it's as simple as adding a sed to replace false with true in `config/environments/production.rb`.

This sounds fine to me. Or using an `ENV` var with fallback to `false`.

Edit, the following worked well for me:
```patch
diff --git Dockerfile Dockerfile
index 2182af555..67679aa87 100644
--- Dockerfile
+++ Dockerfile
@@ -86,6 +86,7 @@ COPY --from=builder --chown=1001:0 ${HOME}/Gemfile.lock ${HOME}/Gemfile.lock
 COPY --from=builder --chown=1001:0 ${HOME}/vendor/ruby ${HOME}/vendor/ruby
 COPY --from=builder --chown=1001:0 ${HOME}/public ${HOME}/public
 RUN rm -rf bundler.d/nulldb.rb bin/spring
+RUN sed -i '/config.public_file_server.enabled/ s/false/true/' config/environments/production.rb
 
 RUN date -u > BUILD_TIME
 
```

evgeni commented on Dockerfile:
```diff
@@ -1,17 +1,14 @@
-  # needed for VNC/SPICE websockets
-  python2-numpy && \
-  microdnf clean all
+  dnf module enable ruby:${RUBY_VERSION} nodejs:$NODEJS_VERSION -y && \
```
```suggestion
  dnf module enable ruby:${RUBY_VERSION} nodejs:${NODEJS_VERSION} -y && \
```

evgeni commented:
> Additionally, since [20f4804](https://github.com/theforeman/foreman/commit/20f4804784c622851fd7beade233797c95b1304b) websockify is unbundled and the container now no longer has it.

We could add nightly foreman repos, and install it from there?

evgeni commented on Dockerfile:
```diff
@@ -33,22 +30,25 @@ ENV FOREMAN_APIPIE_LANGS=en
-RUN echo gem '"tzinfo-data"' > bundler.d/container.rb
-RUN bundle install --without "${BUNDLER_SKIPPED_GROUPS}" \
-    --binstubs --clean --path vendor --jobs=5 --retry=3 && \
+RUN bundle config set --local without "${BUNDLER_SKIPPED_GROUPS}" && \
+  bundle config set --local clean true && \
+  bundle config set --local path vendor && \
+  bundle config set --local jobs 5 && \
+  bundle config set --local retry 3
```
I wonder whether those should be `bundle config` calls vs environment variables, but this is probably not too relevant for the outcome of this PR

ekohl commented:
> We could add nightly foreman repos, and install it from there?

It's in EPEL. We might as well use that.

ekohl commented:
> This sounds fine to me. Or using an `ENV` var with fallback to `false`.

I now implemented an env var which was already set in the container but then unused.

evgeni commented on config/environments/production.rb:
```diff
@@ -16,7 +16,7 @@
 
   # Disable Rails's static asset server (Apache or nginx will already do this).
-  config.public_file_server.enabled = false
+  config.public_file_server.enabled = ENV.fetch('RAILS_SERVE_STATIC_FILES', false) == 'true'
```
This should be documented in https://github.com/theforeman/theforeman.org/blob/gh-pages/_includes/manuals/nightly/3.4_install_from_source.md

evgeni commented:
(it doesn't solve the websockify part yet, but IMHO still worth a merge)

jvstein commented:
I found my way here via [this post](https://community.theforeman.org/t/actioncontroller-routingerror-after-installing-foreman-from-source/31452/3) because the docker image isn't currently working without a change to make `config.public_file_server.enabled = true`.

ekohl commented:
I forgot about this because of release priority and then the holidays, but remembered it today. Rebased this now and addressed one trivial comment. Marking as ready for review.
